docs/source/settings.rst
Settings
========

A settings file contains all the configuration of your KalaPy application. A
settings file is just a normal Python module with module-level variables.

There are two kind of settings file, `project settings` located in the top-level
project directory and `package settings` located in the package directory.


.. _project-settings:

Project Settings
----------------

Here is the list of available project settings.


DATABASE_ENGINE
+++++++++++++++

Default::

    DATABASE_ENGINE = "sqlite3"

The database backend engine. Possible value can be either "sqlite3",
"postgresql", "mysql" or "gae".


DATABASE_NAME
+++++++++++++

Default::

    DATABASE_NAME = ""

The name of the database. For sqlite3, use full path to the sqlite3 database file.

DATABASE_USER
+++++++++++++

Default::

    DATABASE_USER = ""

The database username (must have rights to create database tables). Keep empty
for sqlite3 and gae.

DATABASE_PASSWORD
+++++++++++++++++

Default::

    DATABASE_PASSWORD = ""

The database password. Keep empty for sqlite3 and gae.

DATABASE_HOST
+++++++++++++

Default::

    DATABASE_HOST = ""

The database host. Keep empty for sqlite3 and gae.

DATABASE_PORT
+++++++++++++

Default::

    DATABASE_PORT = ""

The database port. Keep empty for sqlite3 and gae.

DATABASE_OPTIONS
++++++++++++++++

Default::

    DATABASE_OPTIONS = {}

Database specific options.

USE_I18N
++++++++

Default::

    USE_I18N = True

Enable/Disable internationalization support.

DEFAULT_LOCALE
++++++++++++++

Default::

    DEFAULT_LOCALE = "en_US"

Default locale to be used for internationalization and localization.

DEFAULT_TIMEZONE
++++++++++++++++

Default::

    DEFAULT_TIMEZONE = "UTC"

Default timezone, UTC recommended.

MIDDLEWARE_CLASSES
++++++++++++++++++

Default::

    MIDDLEWARE_CLASSES = (
        'kalapy.contrib.sessions.SessionMiddleware',
    )

List of Middleware classes.

SESSION_ENGINE
++++++++++++++

Default::

    SESSION_ENGINE = "memory"

Session storage engine (memory, memcached or database).

SESSION_OPTIONS
+++++++++++++++

Default::

    SESSION_OPTIONS = {
        'memcached_servers': [],
    }

Session related options. For example ``memcached_servers``. For memcached
engine, provide list of memcached servers. If DATABASE_ENGINE is set to
'gae' this option will be ignored.

SESSION_COOKIE
++++++++++++++

Default::

    SESSION_COOKIE = {
        'name': 'session_id',
        'age': 60 * 60 * 24 * 7 * 2,
        'domain': None,
        'path': '/'
    }

Session cookie options.

LOGGING
+++++++

Default::

    LOGGING = {
        'level': 'DEBUG',
        'format': '[%(asctime)s] %(levelname)s:%(name)s:%(message)s',
        'logfile': None,
    }

Settings for logging. The ``level`` can be one of ``'INFO'``, ``'DEBUG'`` or
``'ERROR'``. If ``logfile`` is not provided, `stderr` will be assumed.


STATIC_LINKS
++++++++++++

Default::

    STATIC_LINKS = {
        '/favicon.ico': 'static/favicon.ico',
        '/robots.txt': 'static/robots.txt',
    }

Override the static links. Useful to provide favicon.ico or robots.txt. You can
also provide static directory links to override original static dirs. Paths should
be absolute path or relative to the project directory. For example::

    STATIC_LINKS = {
        '/static': '/path/to/alternative/static/dir',
        '/favicon.ico': 'static/favicon.ico',
        '/foo/static': ('/path/to/foo/static1', '/path/to/foo/static2'),
    }

If you want to specify fallback directories, list them in tuple.

INSTALLED_PACKAGES
++++++++++++++++++

Default::

    INSTALLED_PACKAGES = (
        'kalapy.contrib.sessions',
    )

List of installed packages.

.. _package-settings:

Package Settings
----------------

Here is the list of available package settings.

NAME
++++

Default::

    NAME = "package_name"

The name of the package.

DESCRIPTION
+++++++++++

Default::

    DESCRIPTION = """
    """

Package description.

VERSION
+++++++

Default::

    VERSION = "1.0"

Package version string.

EXTENDS
+++++++

Default::

    EXTENDS = None

The name of the package that is extended by this package. In that case this
package is considered an addon package and resources provided by this package
will be served as the resources of extending package.

DEPENDS
+++++++

Default::

    DEPENDS = None

List of other packages this package depends on.

SUBMOUNT
++++++++

Default::

    SUBMOUNT = None

Submount to be used to mount this package. For example, ``'/wiki'``, ``'/blog'``
etc. Ignored for addon packages.

.. warning::

    As project settings file may contain sensitive information like database
    password, you should limit access to it. For example, change file permission
    etc.


CHANGES
KalaPy Changelog
================

Here you can see the full list of changes between each KalaPy release.

Version 0.4
-----------

Released on 5th August, 2010.

New features
    - Added settings.STATIC_LINKS to override static dirs & links
    - Implemented addon package feature.
    - Implemented logging support.
    - Integrated ipython support (shell command).
    - Added `wiki_extended` example package.
    - Translate Jinja2 templates with easy (see docs).

Refactoring
    - The web module is splited into several smaller modules.
    - Application class is no longer Package subclass
    - ModelCache and Loader have been merged into an object Pool
    - Reimplemented context locals
    - Improved admin commands (help, colorize output, error handling)
    - Lots of new tests
    - Improved documentation

Bugs fixings
    - Fixed early sending of signal `request-finished`.
    - Improved signals api.
    - Fixed GAE issue (sys.path reset, while using `gae runserver`).
    - Fixed a regression (db.String validation).

Version 0.3.1
-------------

Bugfix release, released on 10th July, 2010.

- Fixed GAE issue, convert db.Text values to datastore_types.Text to overcome
  with size contraint.
- The Text field is no longer a subclass of String field (size contraint issue).
- Allow database engines to use custom type convertors.

Version 0.3
-----------

Released on 9th July, 2010.

- Renamed kalapy-admin.py to kalapy-quickstart.py
- Implemented memcached session storage engine.
- Added validator argument to Field constructor.
- Remove unique(), use __unique__ attribute instead.
- Removed validate() decorator, use method name with 'validate_' prefix.
- Removed meta() decorator, don't allow changing meta information.
- Implemented MySQL support.
- Implemented db.run_in_transaction (useful with GAE).
- Improved ManyToMany field (accepts cascade argument)
- Added `gae` admin command.
- Projects are no longer a python packages (fixes GAE issue).
- The example application is now GAE compatible.
- Implemented Google App Engine support
- Refactored db.Query interface
    - Now follows GAE style filter syntax
    - Added db.Q class to support ORed query
    - Added db.QSet to implement engine specific query
- Referance data_type is now engine specific.
- Fixed i18n (loading translation catalogs)
- Fixed Model.set_dirty issue
- Fixed test api (early initialization of Application instance)
- Removed support for auto-jsonify of dict results
- Added more unit tests
- Reimplemented test loader utils
- Test name format now follows 'package:test_name' format
- Added client attribute to TestCase

Version 0.2.1
-------------

Released on 16th May, 2010.

- Fixed distribution packaging issues


Version 0.2
-----------

Released on 15th May, 2010.

- Project has been renamed to KalaPy
- Added a sphinx theme, peacock, for documentation
- Added fr_FR, de_DE and es_ES locale (not translated though)
- Translated all the string instances in the code
- Implemented full i18n support using babel and pytz
- Implemented Babel commands to work with message catalogs
- The get_model now follows same 'package:name' format as web api does
- Improved the pending dependency resolution for Reference fields


Version 0.1
-----------

Released on 3rd May, 2010.

First public preview release.


example/settings.py
# Settings for example project

# Name of the project
PROJECT_NAME = "example"

# Version of the project
PROJECT_VERSION = "1.0"

# Enable/Disable debug mode
DEBUG = True

# Database backend engine
# Possible value can be either sqlite3, postgresql, mysql, gae
DATABASE_ENGINE = "sqlite3"

# Database name
# For sqlite3 use path to the sqlite3 database file
# For gae keep empty
DATABASE_NAME = "example.db"

# Database user (must have rights to create database tables)
# Keep empty for sqlite3 and gae
DATABASE_USER = ""

# Database password
# Keep empty for sqlite3 and gae
DATABASE_PASSWORD = ""

# Database host
# Keep empty for sqlite3 and gae
DATABASE_HOST = ""

# Database port
# Keep empty for sqlite3 and gae
DATABASE_PORT = ""

# Database specific options
DATABASE_OPTIONS = {
}

# Enable/Disable internationalization support
USE_I18N = True

# List of Middleware classes
MIDDLEWARE_CLASSES = (
    'kalapy.contrib.sessions.SessionMiddleware',
)

# Session storage engine (memory, memcached or database)
SESSION_ENGINE = "memory"

# Session related options, for example memcached servers.
#
# For memcached engine, provide list of memcached servers.
# If DATABASE_ENGINE is set to 'gae' this option will be ignored.
SESSION_OPTIONS = {
    'memcached_servers': [],
}

# Session cookie options
SESSION_COOKIE = {
    'name': 'session_id',
    'age': 60 * 60 * 24 * 7 * 2,
    'domain': None,
    'path': '/'
}

# Settings for logging. Ignored if DATABASE_ENGINE is 'gae'.
#
# The level can be one of 'INFO', 'DEBUG' or 'ERROR'
# If logfile is not provided, stderr will be assumed.
LOGGING = {
    'level': 'DEBUG',
    'format': '[%(asctime)s] %(levelname)s:%(name)s:%(message)s',
    'logfile': None,
}

# Some static links. Useful to provide favicon.ico or robots.txt. You can
# also provide static directory links to override original static dirs. The
# paths should be absolute path or relative to the project directory. If you
# want to specify fallback directories, list them in tuple.
#
# For example::
#
#    STATIC_LINKS = {
#        '/static': '/path/to/alternative/static/dir',
#        '/favicon.ico': 'static/favicon.ico',
#        '/foo/static': ('/path/to/foo/static1', '/path/to/foo/static2'),
#    }
#
STATIC_LINKS = {
    '/favicon.ico': 'static/favicon.ico',
    '/robots.txt': 'static/robots.txt',
}

# List of installed packages
INSTALLED_PACKAGES = (
    'kalapy.contrib.sessions',
    'main',
    'wiki',
    'wiki_extended',
)
